src/System.Diagnostics.Debug/tests/DebugTests.cs

```diff
@@ -7,306 +7,40 @@
-            Trace.Listeners.Clear();
-            VerifyLogged(() => Debug.Write("pizza"), string.Empty); 
-            VerifyLogged(() => Trace.Write("pizza"), string.Empty);
-            Trace.Refresh();
-        }
-
-        [Fact]
-        public void Asserts()
-        {
-            VerifyLogged(() => Debug.Assert(true), "");
-            VerifyLogged(() => Debug.Assert(true, "assert passed"), "");
-            VerifyLogged(() => Debug.Assert(true, "assert passed", "nothing is wrong"), "");
-            VerifyLogged(() => Debug.Assert(true, "assert passed", "nothing is wrong {0} {1}", 'a', 'b'), "");
-
-            VerifyAssert(() => Debug.Assert(false), "");
-            VerifyAssert(() => Debug.Assert(false, "assert passed"), "assert passed");
-            VerifyAssert(() => Debug.Assert(false, "assert passed", "nothing is wrong"), "assert passed", "nothing is wrong");
-            VerifyAssert(() => Debug.Assert(false, "assert passed", "nothing is wrong {0} {1}", 'a', 'b'), "assert passed", "nothing is wrong a b");      
-        }
-
-        [Fact]
-        public void Fail()
-        {
-            VerifyAssert(() => Debug.Fail("something bad happened"), "something bad happened");
-            VerifyAssert(() => Debug.Fail("something bad happened", "really really bad"), "something bad happened", "really really bad");        
-        }
-
-        [Fact]
-        public void Write()
-        {
-            VerifyLogged(() => Debug.Write(5), "5");
-            VerifyLogged(() => Debug.Write((string)null), "");
-            VerifyLogged(() => Debug.Write((object)null), "");
-            VerifyLogged(() => Debug.Write(5, "category"), "category:5");
-            VerifyLogged(() => Debug.Write((object)null, "category"), "category:");
-            VerifyLogged(() => Debug.Write("logged"), "logged");
-            VerifyLogged(() => Debug.Write("logged", "category"), "category:logged");
-            VerifyLogged(() => Debug.Write("logged", (string)null), "logged");
-
-            string longString = new string('d', 8192);
-            VerifyLogged(() => Debug.Write(longString), longString);
-        }
+        protected abstract bool DebugUsesTraceListeners { get; }
+        protected static readonly DebugProvider _debugOnlyProvider;
+        protected static readonly DebugProvider _debugTraceProvider;
 
-        [Fact]
-        public void Print()
+        static DebugTests()
         {
-            VerifyLogged(() => Debug.Print("logged"), "logged");
-            VerifyLogged(() => Debug.Print("logged {0}", 5), "logged 5");
+            FieldInfo fieldInfo = typeof(Debug).GetField("s_provider", BindingFlags.Static | BindingFlags.NonPublic);
+            _debugOnlyProvider = (DebugProvider)fieldInfo.GetValue(null);
+            // Triggers code to wire up TraceListeners with Debug
+            Assert.Equal(1, Trace.Listeners.Count);
+            _debugTraceProvider = (DebugProvider)fieldInfo.GetValue(null);
```

Comment:
Maybe assert that the types of the two providers are different to make sure that this is working as expected?

`Assert.NotEqual(_debugOnlyProvider.GetType(), _debugTraceProvider.GetType());`